### PR TITLE
Proposal: Update examples to utilize a new format.

### PR DESCRIPTION
**Pull request by srush**
This PR creates a new example coding style for the pytorch code. 

* Uses pytorch-lightning for the underlying training. 
* Separates out the base transformer loading from the individual training. 
* Moves each individual example to its own directory. 
* Move the code in the readme to bash scripts. 

The only two new files are `run_pl_ner.py` and `transformers_base.py`.

The goal is to keep the same format as the original command-line. Most of the argument names are preserved. I have verified that for NER the results of the same on GPU. 

There are several nice benefits of lightning -> somewhat nicer logging and library integration (e.g. wandb), auto-checkpointing. Mostly the goal though is code readability with identical functionality. 

Todo: 
* make sure that the output file format is identical. 
* print test results after training. 
* test multi-gpu and apex (in theory these should work)

### CODE REVIEW

File: examples/transformer_base.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+    def test_step(self, batch, batch_nb):
+        return self.validation_step(batch, batch_nb)
+
+    def test_end(self, outputs):
+        return self.validation_end(outputs)
+
+    @pl.data_loader
+    def train_dataloader(self):
+        args = self.hparams
```
Comment (sshleifer) [inline examples/transformer_base.py]: (nit) I don't understand purpose of this line.

File: examples/transformer_base.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    parser.add_argument("--seed", type=int, default=42, help="random seed for initialization")
+
+def generic_train(model, args):
+    # init model
+    set_seed(args)
+
+    # Setup distant debugging if needed
+    if args.server_ip and args.server_port:
+        # Distant debugging - see https://code.visualstudio.com/docs/python/debugging#_attach-to-a-local-script
+        import ptvsd
```
Comment (sshleifer) [inline examples/transformer_base.py]: is this a pl dependency? I didn't get it from my `pip install .["dev"]`

File: examples/ner/run_pl_ner.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    all_label_ids = torch.tensor([f.label_ids for f in features], dtype=torch.long)
+
+    dataset = TensorDataset(all_input_ids, all_input_mask, all_segment_ids, all_label_ids)
+    return dataset
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    add_generic_args(parser, os.getcwd())
+    parser = NERTransformer.add_model_specific_args(parser, os.getcwd())
```
Comment (sshleifer) [inline examples/ner/run_pl_ner.py]: This is cool!

**Review by sshleifer — COMMENTED** [review]
LGTM! Looks like we can also move things over one at a time without horrible intermediate states. 

We  should also explicitly drop python 3.5 support and tell circleci to run tests in 3.6. 
We (happy to do it) should probably also add an ner test to `test_examples.py`

File: examples/transformer_base.py (hunk 2 above)
Comment (srush) [reply examples/transformer_base.py]: No this was in the base file I was porting. I will add it as an dev dependency.

File: examples/transformer_base.py (hunk 1 above)
Comment (srush) [reply examples/transformer_base.py]: I'll switch to hparams. (old code had args)

File: examples/ner/run_pl_ner.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
+import argparse
```
Comment (williamFalcon) [inline examples/ner/run_pl_ner.py]: this looks great! looks like you guys did all the  right things

**Comment by stefan-it** [comment]
Hi @srush, thanks for this PR :heart:  Can't wait to test it!

One suggestion/RFC: could we rename it to something like `token_classification` instead of `ner`. I know PoS tagging is not really covered in recent papers, but I always test new models for this task with the "identical" implementation 😅 This requires only a little modification in the code: we then should report accuracy as well.

But I will be totally fine with `ner` here!

**Comment by srush** [comment]
Token Classification sounds good to me. That is consistent with the internal naming.
